### DISCUSSION TITLE
How to fine-tune a .gguf model?

### DISCUSSION THREAD

**Discussion by SeanZhang7** [ggml-org/llama.cpp#7792] [Q&A]
May I ask How could I fine-tune a .gguf model? I saw some discussions about converting it into pytorch form. But this method seems not infeasible.

**Comment by chrisalbertson** [comment]
Total beginner here but, it seems to me what you do is apply an LoRA adaper to the .gguf file and llama.cpp does the work of applying it to the model in real time.

Can you then save the adapted model?   I've not figured that out yet.

**Reply by SeanZhang7** [reply]
I convert a model from HF to .gguf, and I try to fine-tune that gguf model. I haven't figured out hoe to do that. 
Do you mean I can fine-tune it by apply a LoRA adaper to the .gguf file?

**Reply by adewalen06-alt** [reply]
Yh

**Comment by karambaso** [comment]
There is the example here: [llama.cpp/examples/training](https://github.com/ggml-org/llama.cpp/tree/master/examples/training)

Content:

This directory contains examples related to language model training using llama.cpp/GGML. So far finetuning is technically functional (for FP32 models and limited hardware setups) but the code is very much WIP. Finetuning of Stories 260K and LLaMA 3.2 1b seems to work with 24 GB of memory. For CPU training, compile llama.cpp without any additional backends such as CUDA. For CUDA training, use the maximum number of GPU layers.

Proof of concept:

export model_name=llama_3.2-1b && export quantization=f32
./build/bin/llama-finetune --file wikitext-2-raw/wiki.test.raw -ngl 999 --model models/${model_name}-${quantization}.gguf -c 512 -b 512 -ub 512
./build/bin/llama-perplexity --file wikitext-2-raw/wiki.test.raw -ngl 999 --model finetuned-model.gguf

The perplexity value of the finetuned model should be lower after training on the test set for 2 epochs.
